src/main/java/de/dennisguse/opentracks/viewmodels/IntervalStatisticsModel.java

```diff
@@ -1,81 +1,123 @@
     private Distance distanceInterval;
+    private final Distance minGPSDistance;
+    private final ContentResolver contentResolver;
+    private ContentObserver trackPointsTableObserver;
+    private TrackPoint.Id lastTrackPointId;
+
+    private final Executor executor = Executors.newSingleThreadExecutor();
+    private HandlerThread handlerThread;
+    private Handler handler;
 
     public IntervalStatisticsModel(@NonNull Application application) {
         super(application);
+        minGPSDistance = PreferencesUtils.getRecordingDistanceInterval(PreferencesUtils.getSharedPreferences(application), application);
+        contentResolver = getApplication().getContentResolver();
     }
 
-    public MutableLiveData<List<IntervalStatistics.Interval>> getIntervalStats(boolean metricUnits, @Nullable IntervalOption interval) {
-        synchronized (trackPoints) {
-            if (intervalsLiveData == null) {
-                if (interval == null) {
-                    interval = IntervalOption.OPTION_1;
-                }
+    @Override
+    protected void onCleared() {
+        super.onCleared();
+        if (trackPointsTableObserver != null) {
+            contentResolver.unregisterContentObserver(trackPointsTableObserver);
+        }
+        if (handlerThread != null) {
+            handlerThread.getLooper().quit();
+            handlerThread = null;
+        }
+        handler = null;
+    }
 
-                intervalsLiveData = new MutableLiveData<>();
-                distanceInterval = interval.getDistance(metricUnits);
-                loadIntervalStatistics();
+    public MutableLiveData<List<IntervalStatistics.Interval>> getIntervalStats(Track.Id trackId, boolean metricUnits, @Nullable IntervalOption interval) {
+        if (intervalsLiveData == null) {
+            if (interval == null) {
+                interval = IntervalOption.OPTION_1;
             }
-            return intervalsLiveData;
+
+            intervalsLiveData = new MutableLiveData<>();
+            distanceInterval = interval.getDistance(metricUnits);
+            intervalStatistics = new IntervalStatistics(distanceInterval, minGPSDistance);
+
+            loadIntervalStatistics(trackId);
         }
-    }
 
-    private void loadIntervalStatistics() {
-        IntervalStatistics intervalStatistics = new IntervalStatistics(trackPoints, distanceInterval);
-        intervalsLiveData.postValue(intervalStatistics.getIntervalList());
+        registerTrackPointsObserver(trackId);
+
+        return intervalsLiveData;
     }
 
-    public void add(TrackPoint trackPoint) {
-        synchronized (trackPoints) {
-            trackPoints.add(trackPoint);
-        }
+    private void loadIntervalStatistics(Track.Id trackId) {
+        executor.execute(() -> {
+            ContentProviderUtils contentProviderUtils = new ContentProviderUtils(getApplication());
+            List<TrackPoint> trackPoints = contentProviderUtils.getTrackPoints(trackId, lastTrackPointId);
```

Comment:
That is a memory intensive operation; it would be better to pass use the cursor/TrackPointIterator and initialize the intervalStatistics from there.

Reply:
You're right.
I'm surprised how quickly you see all these stuffs!!! :open_mouth: